### PR TITLE
issue_64_fix

### PR DESCRIPTION
**Pull request by scrichar**
Fixes https://github.com/graphite-ng/carbon-relay-ng/issues/64 for our use

### CODE REVIEW

**Comment by scrichar** [comment]
Sorry meant to merge this into my forked repo

**Comment by scrichar** [comment]
Have tested locally, this fixes https://github.com/graphite-ng/carbon-relay-ng/issues/64

File: destination/destination.go [hunk 1]
```diff
@@ -49,6 +50,7 @@ type Destination struct {
 	SpoolSyncPeriod      time.Duration
 	SpoolSleep           time.Duration // how long to wait between stores to spool
 	UnspoolSleep         time.Duration // how long to wait between loads from spool
+	RouteName            string        // how long to wait between loads from spool
```
Comment (Dieterbe) [inline destination/destination.go]: this comment doesn't make any sense

**Comment by Dieterbe** [comment]
this PR looks pretty good. have to analyze a bit deeper though.
is it backwards compatible? do users need to specify a spool name or will it default to what it was before?

File: destination/destination.go (hunk 1 above)
Comment (scrichar) [reply destination/destination.go]: yeah, my bad, copied from line above, will remove that

**Comment by scrichar** [comment]
You do not need to specify a spool name, the spoolName would now be derived from a combination of the routeName and the destination/port. Thus providing a way to have unique spool names across different routes with the same destination/port. When the carbon-relay starts up after an upgrade with this fix, the relay would create brand new spools, and essentially any spool that existed before the upgrade would be orphaned. Perhaps I could add a note in the ReadME, for example: "When upgrading to a release > v0.9.4, verify any critical data in the spools is emptied, or the data is not critical and can be orphaned as spool naming convention has changed."

**Comment by Dieterbe** [comment]
sounds good. don't worry about the readme. we can put that in the release notes once we make a release

File: util/util.go [hunk 2]
```diff
@@ -9,3 +9,11 @@ func AddrToPath(s string) string {
 	s = strings.Replace(s, ":", "_", -1)
 	return strings.Replace(s, "/", "", -1)
 }
+
+// some.host:2003, kafkaRoute -> kafkaRoute_some_host_2003
+// http://some.host:8080, kafkaRoute -> kafkaRoute_http_some_host_8080
+func SpoolKeyAddrToPath(s string, routeName string) string {
+	s = AddrToPath(s)
+	sList := []string{routeName, s}
+	return strings.Join(sList, "_")
```
Comment (Dieterbe) [inline util/util.go]: this is unnecessarily complicated
can just `return routeName + "_" + AddrToPath(s)`

**Comment by Dieterbe** [comment]
note also that we currently set `cleanAddr` and base a lot of stats on that.
looks like the stats suffer the same problem as spool names (multiple dests with same addr clobbering each other's stats).
ditto for log calls, which use the dest.Addr property

so my suggestion is , rename the spoolKey property to simply 'Key', and use it for all stats and log calls.
remove the cleanAddr property.

this way we solve the problem for spool files as well as log messages as well as stats.

File: util/util.go (hunk 2 above)
Comment (scrichar) [reply util/util.go]: agreed

**Comment by scrichar** [comment]
@Dieterbe , I have tested these changes locally, things are looking good on my fork. Logs and stats are now using the new property "key". cleanAddr is removed. Spooling is now using the combination of the route name and the destination/port combo. Can you review?

**Comment by Dieterbe** [comment]
looks pretty good.
I did squash all your commits into one and applied some cleanups. see 7a664ab7893c48c479170049b33f8a2cec6062f1

thanks a lot for your contribution, this was a longstanding pain point and i'm glad you stepped up :)

**Pull request closed without merging**